lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,612 @@
+			err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+		}
+		if (err != 0) {
+			return (err);
+		}
+		return (EFBIG);
+	}
+
+	return (0);
+}
+
+static int
+libufzs_inode_lookup_kvattr_zap(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    const char *name, uint64_t *xattr_zap_obj)
+{
+	objset_t *os = sa_hdl->sa_os;
+	int err = sa_lookup(sa_hdl, sa_tbl[UZFS_ZXATTR],
+	    xattr_zap_obj, sizeof (*xattr_zap_obj));
+	if (err != ENOENT && err != 0) {
+		return (err);
+	}
+	if (err == 0) {
+		err = zap_contains(os, *xattr_zap_obj, name);
+	}
+	return (err);
+}
+
+static void
+libuzfs_inode_save_kvattr(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    boolean_t existed_in_sa, boolean_t sa_space_enough, char *xattr_sa_data,
+    uint64_t xattr_sa_size, boolean_t existed_in_zap, uint64_t xattr_zap_obj,
+    uint64_t zap_entries_count, const char *name, const char *value,
+    size_t size, dmu_tx_t *tx)
+{
+	objset_t *os = sa_hdl->sa_os;
+	if (existed_in_sa || sa_space_enough) {
+		if (xattr_sa_size == 0) {
+			VERIFY0(sa_remove(sa_hdl, sa_tbl[UZFS_XATTR], tx));
+		} else {
+			VERIFY0(sa_update(sa_hdl, sa_tbl[UZFS_XATTR],
+			    xattr_sa_data, xattr_sa_size, tx));
+		}
+	}
+
+	if (existed_in_zap && sa_space_enough) {
+		ASSERT(zap_entries_count >= 1);
+		if (zap_entries_count == 1) {
+			VERIFY0(zap_destroy(os, xattr_zap_obj, tx));
+			VERIFY0(sa_remove(sa_hdl, sa_tbl[UZFS_ZXATTR], tx));
+		} else {
+			VERIFY0(zap_remove(os, xattr_zap_obj, name, tx));
+		}
+	}
+
+	if (!sa_space_enough) {
+		if (xattr_zap_obj == DMU_NEW_OBJECT) {
+			int dnodesize = dmu_objset_dnodesize(os);
+			xattr_zap_obj = zap_create_dnsize(os,
+			    DMU_OT_DIRECTORY_CONTENTS, DMU_OT_NONE, 0,
+			    dnodesize, tx);
+			VERIFY0(sa_update(sa_hdl, sa_tbl[UZFS_ZXATTR],
+			    &xattr_zap_obj, sizeof (xattr_zap_obj), tx));
+		}
+		VERIFY0(zap_update(os, xattr_zap_obj, name,
+		    1, size, value, tx));
+	}
+}
+
+int
+libuzfs_inode_set_kvattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const char *name, const char *value, uint64_t size,
+    int flags, uint64_t *txg)
+{
+	// only support value size <= 8K
+	ASSERT3U(size, <=, ZAP_MAXVALUELEN);
+
+	sa_handle_t *sa_hdl = NULL;
+	objset_t *os = dhp->os;
+	int err = sa_handle_get(os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	nvlist_t *nvl;
+	sa_attr_type_t *sa_tbl = dhp->uzfs_attr_table;
+	err = libuzfs_get_nvlist_from_handle(sa_tbl, &nvl, sa_hdl);
+	if (err == ENOENT) {
+		err = nvlist_alloc(&nvl, NV_UNIQUE_NAME, KM_SLEEP);
+	}
+	if (err != 0) {
+		goto out_handle;
+	}
+
+	boolean_t existed_in_sa = nvlist_exists(nvl, name);
+	boolean_t sa_space_enough = B_TRUE;
+	size_t xattr_sa_size = 0;
+	err = libuzfs_kvattr_set_nvlist(nvl, name, value, size, &xattr_sa_size);
+	if (err == EFBIG) {
+		sa_space_enough = B_FALSE;
+	} else if (err != 0) {
+		goto out_nvl;
+		return (err);
+	}
+
+	char *xattr_sa_data = NULL;
+	if (existed_in_sa || sa_space_enough) {
+		if (xattr_sa_size > 0) {
```

Comment:
if sa space is enough, xattr_sa_size > 0 is always true, isn't it?

Reply:
this is true, but if existed in sa and sa space not enough, we need to delete an entry from sa, which may cause xattr_sa_size to be 0

Reply:
so this check could be refined as this to make it more clear:

if (sa_space_enough || (existed_in_sa && xattr_sa_size > 0)) {
    // pack nvlist
}